### PR TITLE
Fix boom-3633

### PR DESCRIPTION
**Pull request by marionf**
Fix issue: https://github.com/PrestaShop/PrestaShop/issues/9924

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
To be reviewed by someone else, I allowed myself to push on the branch as no fork has been created.

File: ps_mainmenu.php [hunk 1]
```diff
@@ -942,20 +945,39 @@ public function hookCategoryUpdate($params)
...
 
     protected function clearMenuCache()
     {
-        $dir = $this->getCacheDirectory();
-
-        if (!is_dir($dir)) {
-            return;
-        }
+        $this->cleanMenuCacheDirectory(
+            $this->getCacheDirectory()
```
Comment (jolelievre) [inline ps_mainmenu.php]: Maybe you should clear all the cache here? Without considering the customer groups
```php
$this->cleanMenuCacheDirectory(_PS_CACHE_DIR_ . 'ps_mainmenu');
```
(assuming `cleanMenuCacheDirectory` is recursive in the folders)

File: ps_mainmenu.php [hunk 2]
```diff
@@ -712,10 +712,13 @@ protected function generateCategoriesMenu($categories, $is_children = 0)
 
         foreach ($categories as $key => $category) {
             $node = $this->makeNode([]);
-
             if ($category['level_depth'] > 1) {
                 $cat = new Category($category['id_category']);
                 $link = $cat->getLink();
+                // Check if customer is set and check access
+                if (isset($this->context->customer) && !$cat->checkAccess($this->context->customer->id)) {
```
Comment (PierreRambaud) [inline ps_mainmenu.php]:
```suggestion
                if (isset($this->context->customer->id) && !$cat->checkAccess($this->context->customer->id)) {
```

Comment (marionf) [reply ps_mainmenu.php]: Done

File: ps_mainmenu.php (hunk 1 above)
Comment (marionf) [reply ps_mainmenu.php]: Done

**Comment by marionf** [comment]
I have this error when I try to edit and save a category but don't know how to fix it :sweat_smile: 

capture d'écran_1262

If someone know I let you do it, otherwise this PR can be closed

**Comment by jolelievre** [comment]
@marionf you can add:
```php
if (!is_dir($dir)) {
    return;
}
```
at the beginning of `cleanMenuCacheDirectory `

File: ps_mainmenu.php [hunk 3]
```diff
@@ -942,20 +945,37 @@ public function hookCategoryUpdate($params)
...
     {
-        $dir = $this->getCacheDirectory();
-
-        if (!is_dir($dir)) {
-            return;
-        }
+        $this->cleanMenuCacheDirectory(_PS_CACHE_DIR_ . 'ps_mainmenu');
+    }
 
+    private function cleanMenuCacheDirectory(string $dir)
```
Comment (Matt75) [inline ps_mainmenu.php]:
```php
private function cleanMenuCacheDirectory(string $dir)
```
should be
```php
private function cleanMenuCacheDirectory($dir)
```
Scalar type hints are available in PHP 7 only

File: ps_mainmenu.php [hunk 4]
```diff
@@ -942,20 +945,37 @@ public function hookCategoryUpdate($params)
 
     protected function getCacheDirectory()
     {
-        return _PS_CACHE_DIR_ . 'ps_mainmenu';
+        $dir =_PS_CACHE_DIR_ . 'ps_mainmenu';
+
+        if (isset($this->context->customer)) {
+            $groups = $this->context->customer->getGroups();
+            if (count($groups)) {
```
Comment (Matt75) [inline ps_mainmenu.php]:
```php
if (count($groups)) {
```
Should be
```php
if (!empty($groups)) {
```

File: ps_mainmenu.php [hunk 5]
```diff
@@ -712,10 +712,13 @@ protected function generateCategoriesMenu($categories, $is_children = 0)
 
         foreach ($categories as $key => $category) {
             $node = $this->makeNode([]);
-
             if ($category['level_depth'] > 1) {
                 $cat = new Category($category['id_category']);
                 $link = $cat->getLink();
+                // Check if customer is set and check access
+                if (isset($this->context->customer->id) && !$cat->checkAccess($this->context->customer->id)) {
```
Comment (Matt75) [inline ps_mainmenu.php]:
```php
if (isset($this->context->customer->id) && !$cat->checkAccess($this->context->customer->id)) {
```
Should be
```php
if (Validate::isLoadedObject($this->context->customer) && !$cat->checkAccess($this->context->customer->id)) {
```

File: ps_mainmenu.php (hunk 3 above)
Comment (marionf) [reply ps_mainmenu.php]: Done :)

File: ps_mainmenu.php (hunk 4 above)
Comment (marionf) [reply ps_mainmenu.php]: Done :)

File: ps_mainmenu.php (hunk 5 above)
Comment (marionf) [reply ps_mainmenu.php]: Done :)

**Comment by marionf** [comment]
Thank you guys, I added your comments

**Comment by jolelievre** [comment]
Seems good to me, if no one else has any feedback I think it's good to be merged (QA seems to have been approved)

**Comment by PierreRambaud** [comment]
Thanks @marionf

**Pull request merged**